[lldb] Handle diagnostics better around expression evaulation retries

augusto2112 opened the pull request:
The error message produced when expression evaluation parsing failed with automatic retries was not great, and would confuse many users. Change the logic around what diagnostics should be displayed to the user if expression evaluation parsing retry also fails to show the original error message.

rdar://118057664

augusto2112 commented:
@swift-ci test

adrian-prantl reviewed (commented):
I'm generally fine with the approach, I would just like you to double-check that SwiftASTContext::GetScopedDiagnosticConsumer() isn't better suited for the job.

adrian-prantl commented:
An maybe the answer is no

adrian-prantl commented:
Do you have a testcase?

augusto2112 commented:
> I'm generally fine with the approach, I would just like you to double-check that SwiftASTContext::GetScopedDiagnosticConsumer() isn't better suited for the job.

As I understand it, SwiftASTContext::GetScopedDiagnosticConsumer is used to capture diagnostics generated by the swift compiler, in this case we're dealing with the more general lldb diagnostics.

augusto2112 committed:
[lldb] Implement DiagnosticManager::Consume

In some situations it may be useful to have a separate DiagnosticManager
instance, and then later of move the contents of that instance back to
the "main" DiagnosticManager. For example, when silently retrying some
operation with different parameters, depending on whether the retry
succeeded or failed, LLDB may want to present a different set of
diagnostics to the user (the ones generated on the first try vs the
retry). Implement DiagnosticManager::Consume to allow for this use case.

augusto2112 committed:
[lldb] Handle diagnostics better around expression evaulation retries

The error message produced when expression evaluation parsing failed
with automatic retries was not great, and would confuse many users.
Change the logic around what diagnostics should be displayed to the
user if expression evaluation parsing retry also fails to show the
original error message.

rdar://118057664

augusto2112 commented:
@swift-ci test

augusto2112 commented:
@adrian-prantl had to change my approach a little, if we decide this is the way to go I'll upstream the patch that adds `DiagnosticManager::Consume` before merging this one.

augusto2112 commented:
@swift-ci test windows

adrian-prantl commented on lldb/test/API/lang/swift/fixits/TestSwiftFixIts.py:
```diff
@@ -65,21 +65,6 @@ def do_test(self):
         options = lldb.SBExpressionOptions()
         options.SetAutoApplyFixIts(False)
 
-        # First make sure the expressions fail with no fixits:
```
Why are these removed?

augusto2112 commented:
@swift-ci test

adrian-prantl commented on lldb/test/API/lang/swift/private_generic_type/TestSwiftPrivateGenericType.py:
```diff
@@ -151,3 +151,8 @@ def test_private_generic_type(self):
                     error=True)
 
+        # Check that if both binding and not binding the generic type parameters fail, we report 
+        # the "bind generic params" erroe message, as that's the default case that runs first.
```
```suggestion
        # the "bind generic params" error message, as that's the default case that runs first.
```

augusto2112 commented:
@swift-ci test

augusto2112 commented:
@swift-ci test Windows

augusto2112 commented:
@swift-ci test macOS

augusto2112 commented:
@swift-ci test Windows

augusto2112 commented:
@swift-ci test macOS

augusto2112 commented:
@swift-ci test Windows

adrian-prantl commented:
@swift-ci test macos

adrian-prantl commented:
@swift-ci test macos
